### ISSUE TITLE
Webradio runs as root

### ISSUE THREAD

**Issue by HiranChaudhuri** [HiranChaudhuri/Webradio_v4#22]
Why does Webradio need so much privileges? Can it run as normal user?

**Comment by Acer54** [comment]
Webradio was developed at a time, where the usage of GPIOs was only possible with root-privileges.
In the meantime it should not be necessary (I think this change was introduced with Raspbian Jessie or Stretch).
From my point of view it is not necessary any longer to run webradio-py with sudo.

We need a different structure of the files which are executed then. 
At the moment there is a system-wide "systemd-service" which is the trigger for the program-start (or autostart).
So the systemd-service runs with root-privileges. If sudo with the User-Definition (pi) is missing, it will run the program with root-privileges anyway but it is not possible to determine who is logged in at time of execution...

Do you have a minimum-example of a PyQt (GUI) Application running as systemd-service (started or autostarted with su-rights) able to grep the logged in user (with os.path.expand("~") or getpass.getpass())

Was not able to achieve this at my Ubuntu system... only utilizing "systemctl --user" command...

**Comment by HiranChaudhuri** [comment]
So far I see that being member of the group "gpio" should allow access to those pins.
I would have suggested to create a user "webradio" and assign that to the required groups.
However the bigger challenge might be getting access to the X display.

I'll try to figure out how that works.

**Comment by HiranChaudhuri** [comment]
It seems someone managed to use systemd for running a graphical application:
https://www.linuxquestions.org/questions/linux-desktop-74/start-gui-application-from-systemd-unit-service-4175591018/

and it seems someone else was suggested to have systemd startup X, which starts xinit, which starts the windowmanager, which autostarts the GUI application
https://stackoverflow.com/questions/41459360/cannot-launch-gui-programs-from-systemd-service

**Comment by Acer54** [comment]
Hm, your first link, is more or less the same what the current launcher (systemd-service) is doing.

For the second link, we also should have in mind, that it should be possible to install webradio-py as an "application" in a system which might be used in other ways as well, not a stand-alone image. This would be a different approach... when we can offer a pre-configured ISO-Image like other Pi-Projects do....

**Comment by HiranChaudhuri** [comment]
> Hm, your first link, is more or less the same what the current launcher (systemd-service) is doing.
> 
> For the second link, we also should have in mind, that it should be possible to install webradio-py as an "application" in a system which might be used in other ways as well, not a stand-alone image. This would be a different approach... when we can offer a pre-configured ISO-Image like other Pi-Projects do....

I totally agree. However for the first case I'd like to better understand how another user could gain access to the pi user's graphical desktop.

In the second case I am not worried about whether we'd run an appliance (webradio is the only application) or we just integrate into the desktop. You still have the interactive configuration (which I do not like as I would not know how to script it), and in worst case we could offerdifferent debian packages. But I think we still need a good understanding how to do it properly.

One of the first steps from my point of view is to have a good test environment. Do you know of fully scripted Raspberry setups?

**Comment by Acer54** [comment]
This should be a good point for a start : [https://github.com/RPi-Distro/pi-gen](https://github.com/RPi-Distro/pi-gen)
On the other hand, one can create a complete, pre-configured image-file which can be written to a Mico-SSD with dd or etcher.

**Comment by HiranChaudhuri** [comment]
The pi-gen project in combination with https://www.pcsteps.com/1199-raspberry-pi-emulation-for-windows-qemu/ should allow us to fully script a test environment. Then it would be about automatically running tests, which is another chapter. I'll try to find some time for that.

**Comment by HiranChaudhuri** [comment]
For this actual issue, I found out that
* in Raspbian, systemd has a service /etc/systemd/system/display-manager.service that references LightDM. It will run /usr/sbin/lightdm without parameters.
* according to https://github.com/canonical/lightdm the configuration files are
** /usr/share/lightdm/lightdm.conf.d/*.conf
** /etc/lightdm/lightdm.conf.d/*.conf
** /etc/lightdm/lightdm.conf
Somehow lightdb launches lxsession, which again launches Openbox as window manager.

Suitable entry points for direct Webradio startup would be lxsession or Openbox.

**Comment by HiranChaudhuri** [comment]
So I chose the entry point directly in the lxsession. I did this by editing /etc/xdg/lxsession/LXDE-pi/autostart to contain
`@python3 /opt/webradio-py3/webradio.py --touchscreen --fullscreen`

After a reboot webradio starts and is usable. The fullscreen mode shows an offset of +2+2 pixels from the top left corner.

**Comment by HiranChaudhuri** [comment]
I perceive problems when searching for something that does not exist. The dialog box is not closeable.
A check reveals this was not introduced by the missing window manager - the problem also exists when running the code as usual.
